lib/spack/spack/database.py

```diff
@@ -409,31 +409,83 @@ def _read_suppress_error():
                 self._data = {}
 
         transaction = WriteTransaction(
-            self.lock, _read_suppress_error, self._write, _db_lock_timeout)
+            self.lock, _read_suppress_error, self._write, _db_lock_timeout
+        )
 
         with transaction:
             if self._error:
                 tty.warn(
                     "Spack database was corrupt. Will rebuild. Error was:",
-                    str(self._error))
+                    str(self._error)
+                )
                 self._error = None
 
+            # Read first the `spec.yaml` files in the prefixes. They should be
+            # considered authoritative with respect to DB reindexing, as
+            # entries in the DB may be corrupted in a way that still makes
+            # them readable. If we considered DB entries authoritative
+            # instead, we would perpetuate errors over a reindex.
```

Comment:
We *probably* should add an option to propagate only specs from `spec.yaml` files, but that can wait for a separate PR.  I don't think there is an immediate need for it.